src/CSVMsg.chpl

```diff
@@ -197,7 +197,32 @@ module CSVMsg {
             }
         }
 
-        coforall loc in Locales do on loc{
+        // only use 90% of getMemLimit to give ourselves some buffer
+        // these are each per locale measurements
+        const memLim = (getMemLimit() * (.9)):uint;
+        const memUsed = getMemUsed();
+
+        // calculate the amount of memory required to write the csv file all in one go
+        var memRequired: uint = 0;
+        var numStrings = 0;
+        for (cname, dt) in zip(dtypes, datasets) {
+            if dt == "Strings" {
+                var ss = new SegString("", toSegStringSymEntry(st.lookup(cname)));
+                // strings vals are uint(8), so they're only one byte but overcount
+                // because they're not evenly distributed across locales
+                memRequired += (ss.values.a.size * 1.5):int;
+                numStrings += 1;
+            }
+        }
+        // assume pdarrays are 64 bits / 8 bytes (this is an overestimate for bools, but that's fine)
+        memRequired += ((ndsets-numStrings) * row_count * 8);
+        memRequired = (memRequired:real / numLocales):uint;
+
+        // if we can put the entire df in mem (i.e. memLim > (memUsed + memRequired)), then we can do it all in one go.
+        // Otherwise we can only put (memLim - memUsed) in memory at a time, so that's our batchSize
+        const numBatches = if memLim > (memUsed + memRequired) then 1 else ceil(memRequired:real / (memLim - memUsed)):int;
```

Comment:
I don't think the ternary is necessary:
`memLim > (memUsed + memRequired)`  =>
`memLim - memUsed >  memRequired` = >
`1 > memRequired/(memLim - memUsed)` =>
`ceil(memRequired:real / (memLim - memUsed)) == 1`

Reply:
Ah yah! I originally was gonna use a bool to switch between the orig code and the batch. But eventually I changed it to always batch and sometimes only have one batch. It seems I didn't think too hard once I had that change